common/src/main/java/com/github/twitch4j/common/pool/SubscriptionConnectionPool.java

```diff
@@ -0,0 +1,180 @@
+     */
+    public int numSubscriptions() {
+        int n = 0;
+
+        n += saturatedConnections.size() * maxSubscriptionsPerConnection;
+
+        for (Integer i : unsaturatedConnections.values()) {
+            n += i;
+        }
+
+        return n;
+    }
+
+    protected abstract T handleSubscription(C c, S s);
+
+    protected abstract T handleDuplicateSubscription(C c, S s);
+
+    protected abstract U handleUnsubscription(C c, T t);
+
+    protected abstract S getRequestFromSubscription(T t);
+
+    private C getOrCreateConnectionWithHeadroomAndIncrement() {
+        final int max = this.maxSubscriptionsPerConnection;
+
+        // Attempt to find an existing unsaturated connection
+        // Implementation Notes:
+        // With this loop, there is no guarantee regarding entry order over time.
+        // Without unsubscriptions, this algorithm prefers to keep a minimal amount of unsaturated connections (one, if possible without locking)
+        // and fill them completely before moving onto another unsaturated connection.
+        // With unsubscriptions, the behavior is not as deterministic as the elements are not constantly reordered by to subscription count.
+        // Lastly, if multiple threads attempt to make a subscription at the same time and there are no existing unsaturated connections,
+        // this code may create a new connection for each of these threads, due to the lock-free approach. Synchronization would avoid this.
+        AtomicBoolean foundUnsaturated = new AtomicBoolean();
+        for (C connection : unsaturatedConnections.keySet()) {
+            // Try to increment this connection atomically
+            final Integer computed = unsaturatedConnections.compute(connection, (c, n) -> {
+                if (n == null || n + 1 > max)
+                    return null;
+
+                foundUnsaturated.set(true);
+                final int n2 = n + 1;
+                return n2 < max ? n2 : null; // remove from unsaturated if at max capacity
+            });
+
+            if (foundUnsaturated.get()) {
+                // Check if the connection has further headroom or needs to be marked as saturated
+                if (computed == null)
+                    saturatedConnections.add(connection);
+
+                return connection; // found a sufficient existing connection!
+            }
+        }
+
+        // Fallback to creating a new connection (and incrementing that)
+        final C c = createConnection();
+        if (c != null) {
+            // noinspection ConstantConditions (SuperBuilder related)
+            if (1 < max)
+                unsaturatedConnections.putIfAbsent(c, 1);
+            else
+                saturatedConnections.add(c);
+        }
+        return c;
+    }
+
+    private void decrementSubscriptions(C connection) {
+        // Can no longer be saturated
+        saturatedConnections.remove(connection);
```

Comment:
Because there is a short period of time after this removal takes place and `unsaturatedConnections` is not yet updated, it is possible for `getConnections`/`numConnections`/`numSubscriptions` to yield underestimates on concurrent threads as there is no synchronization. 

Scenarios where this could be problematic:
1. calling pool.close at the same time as a saturated connection is being decremented to unsaturated on another thread
2. tracking connection/subscription count in a multi-threaded test where enough `subscribe`s and `unsubscribe`s occur
3. implementing a pool with a max connections count restriction (though one would also need to synchronize the fallback code in `getOrCreateConnectionWithHeadroomAndIncrement` here)

Due to 1, it would be better to provide an overestimate if the lock-free approach were maintained. Here, dispose may just be called twice on the same connection, which is okay. 2 is not reflective of real-world usage and 3 already requires additional synchronization elsewhere